gr-blocks/python/blocks/matrix_interleaver.py

```diff
@@ -0,0 +1,67 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 Caliola Engineering, LLC.
+#
+# This is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 3, or (at your option)
+# any later version.
+#
+# This software is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this software; see the file COPYING.  If not, write to
+# the Free Software Foundation, Inc., 51 Franklin Street,
+# Boston, MA 02110-1301, USA.
+#
+
+from gnuradio import gr, blocks
+
+class matrix_interleaver(gr.hier_block2):
+    def __init__(self, itemsize, rows=1, cols=1, deint=False):
```

Comment:
A couple of suggestions that you can consider here:

* I always get confused about the meaning of rows and columns. Should data be written by rows and read by columns or the other way around? I think that there are different conventions in the literature, so I suggest to document clearly the approach we follow here (specially because it's not so obvious to me by looking at the code below). I've seen this is explained in the GRC block documentation, but probably it's worth to add the same as a comment in the Python code.

* I think that deinterleaving can be seen as the same process as interleaving but exchanging the number of rows and number of columns (i.e., matrix transpose). So we could do without the `deint` parameter. I think this is more a matter of taste, though, and I can see people arguing that they prefer to have a deinterleave option than to have to think about this property.

Reply:
> First thing that I should be said is that I think that this PR is quite useful. The matrix interleaver here serves in more than 90% of the cases. And I like that it includes an example flowgraph and unit test.

I really appreciate this review @daniestevez and the thoughtfulness of the recommendations. 

> 
> I'm not certain if the approach followed here is the best, however. Interleaver codewords usually have hundreds or thousands of elements, so the number of columns and rows of the matrix is usually a few tens or hundreds. For this implementation, this means that the number of ports connecting the deinterleave and interleave blocks can be quite large. 

This block was developed in the context of long streams of contiguous PHY frames you designate "back-to-back". The typical interleaver in my flowgraph has 8 rows & block sizes up to 312 kbits. Most of my testing has been with block sizes under 20 kbits which still means GR is handling around 3 k outputs (thus buffers) for the deinterleave sub-block in matrix deinterleaver mode. I haven't looked at the performance impact of block size scale at all so that may be an issue.

I also haven't looked at the (de)interleave to know if those implementations are particularly optimized, but this matrix (de)interleaver block would automatically benefit from any improvements to those.

> I think that a more efficient implementation could be done in C++, use `set_output_multiple(rows * columns)`, and basically copy the data to the output in the appropriate order by striding through the input samples. The best thing about this PR is that this C++ implementation can be done at a later stage, and only the hierarchical Python flowgraph will need to be replaced. The rest of the PR (test, example, GRC block) could still be re-used.

Agree. I think this block could easily work alongside an arbitrary interleaver in the interim until tests could definitively declare a uniformly superior solution.

Reply:
I don't know why my comment above showed up under this code comment when it was intended to respond to [the comment above](https://github.com/gnuradio/gnuradio/pull/4861#issuecomment-880981384), too much in the interface for me to parse apparently. 

To respond to this code comment, definitely I can add documentation to the python code. I like the deinterleave flag so that it reduces the cognitive load on the developer. Much easier to simply copy the interleaver block and select the flag, than swap arguments I believe. Maybe some GRC-fu to change the appearance of the block when the deint flag is asserted?

Reply:
Regarding GRC-fu, not that I'm aware of, besides what you're already getting by using `hide: none` in the deint option in the YAML file. Another option could be to have two different blocks GRC for interleave and deinterleave, like we have scramble and descrambler, and make both use the same Python block. I slightly prefer a single block with the boolean flag, as you've made.